demos/unity_states/unity/Assets/Scripts/BallStateDataSourcePlatform.cs

```diff
@@ -0,0 +1,224 @@
+using UnityEngine;
+using TP = Teleportal.Client;
+using States = Teleportal.Client.Contract.Properties.States;
+using RSharp;
+
+// Attach to Ball prefab.
+[RequireComponent(typeof(BallStateDataSourceUnity))]
+public class BallStateDataSourcePlatform : MonoBehaviour, IBallStateDataSource
+{
+    private BallStateDataSourceUnity dataSourceUnity;
+
+    // Platform interop types
+    private TP.Baseline baselineMain;
+    private BallContract ballContract;
+    private TP.Object.ObjectHandle ballObject;
+    private BallStates ballStates;
+
+    private States.StateHandle_F32 stateHandlePosX;
+    private States.StateHandle_F32 stateHandlePosY;
+    private States.StateHandle_F32 stateHandlePosZ;
+    private States.StateHandle_I16 stateHandleEulerX;
+    private States.StateHandle_I16 stateHandleEulerY;
+    private States.StateHandle_I16 stateHandleEulerZ;
+    private States.StateHandle_F32 stateHandleScaleX;
+    private States.StateHandle_F32 stateHandleScaleY;
+    private States.StateHandle_F32 stateHandleScaleZ;
+    private States.StateHandle_U64 stateHandleColor;
+
+    void Awake()
+    {
+        this.dataSourceUnity = GetComponent<BallStateDataSourceUnity>();
+        InstantiatePlatformObject();
+    }
+
+    void Update()
+    {
+        ApplyUnityStateToPlatformBaseline();
+        LogCurrentData();
+    }
+
+    private void InstantiatePlatformObject()
+    {
+        this.baselineMain = new TP.Baseline(true);
+        this.ballContract = BallContract.Register(this.baselineMain);
+        this.ballObject = this.ballContract.ObjectCreate(
+            this.baselineMain,
+            this.dataSourceUnity.pos_x, this.dataSourceUnity.pos_y, this.dataSourceUnity.pos_z,
+            this.dataSourceUnity.euler_x, this.dataSourceUnity.euler_y, this.dataSourceUnity.euler_z,
+            this.dataSourceUnity.scale_x, this.dataSourceUnity.scale_y, this.dataSourceUnity.scale_z,
+            this.dataSourceUnity.color
+        );
+
+        ConfigurePlatformObjectStates();
+    }
+
+    private void ConfigurePlatformObjectStates()
+    {
+        this.ballStates = this.ballContract.States;
```

Comment:
this is a shared ref in rust to the contrac'ts states, which are stored in the baseline. Later you get a mut ref to the baseline. I think (?) this is technically violating rust's aliasing guarantees and therefore you should not hold onto this. Just hold the handles instead.

If you have any ideas how to guard against this behavior inside C#, lmk

Reply:
Ah true. Will convert to using the ref only momentarily in order to get the handles from C#.